source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,275 @@ extern "C"
+      dummy;                                    // the numbers of edges cut by the
+      // resulting partition
+
+      // We can not partition n items into more than n parts. METIS will
+      // generate non-sensical output (everything is owned by a single process)
+      // and complain with a message (but won't return an error code!):
+      // ***Cannot bisect a graph with 0 vertices!
+      // ***You are trying to partition a graph into too many parts!
+      nparts = std::min(n, nparts);
+
+      // use default options for METIS
+      idx_t options[METIS_NOPTIONS];
+      METIS_SetDefaultOptions (options);
+
+      // one more nuisance: we have to copy our own data to arrays that store
+      // signed integers :-(
+      std::vector<idx_t> int_rowstart(1);
+      int_rowstart.reserve(sparsity_pattern.n_rows()+1);
+      std::vector<idx_t> int_colnums;
+      int_colnums.reserve(sparsity_pattern.n_nonzero_elements());
+      for (SparsityPattern::size_type row=0; row<sparsity_pattern.n_rows(); ++row)
+        {
+          for (SparsityPattern::iterator col=sparsity_pattern.begin(row);
+               col < sparsity_pattern.end(row); ++col)
+            int_colnums.push_back(col->column());
+          int_rowstart.push_back(int_colnums.size());
+        }
+
+      std::vector<idx_t> int_partition_indices (sparsity_pattern.n_rows());
+
+      // Make use of METIS' error code.
+      int ierr;
+
+      // Select which type of partitioning to create
+
+      // Use recursive if the number of partitions is less than or equal to 8
+      if (nparts <= 8)
+        ierr = METIS_PartGraphRecursive(&n, &ncon, int_rowstart.data(), int_colnums.data(),
+                                        nullptr, nullptr, nullptr,
+                                        &nparts,nullptr,nullptr,options,
+                                        &dummy,int_partition_indices.data());
+
+      // Otherwise use kway
+      else
+        ierr = METIS_PartGraphKway(&n, &ncon, int_rowstart.data(), int_colnums.data(),
+                                   nullptr, nullptr, nullptr,
+                                   &nparts,nullptr,nullptr,options,
+                                   &dummy,int_partition_indices.data());
+
+      // If metis returns normally, an error code METIS_OK=1 is returned from
+      // the above functions (see metish.h)
+      AssertThrow (ierr == 1, ExcMETISError (ierr));
+
+      // now copy back generated indices into the output array
+      std::copy (int_partition_indices.begin(),
+                 int_partition_indices.end(),
+                 partition_indices.begin());
+#endif
+    }
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    //Query functions for partition_zoltan
+    int get_number_of_objects(void *data, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+
+      *ierr = ZOLTAN_OK;
+
+      return graph->n_rows();
+    }
+#endif
```

Comment:
you don't have to put `#endif` here, just put it right before ` void partition_zoltan`. Then you can remove all those `#ifdef` in between.